### PR TITLE
Implicit lock preservation

### PR DESCRIPTION
**Pull request by indirect**
When either `bundle check` is run, or any application requires the `bundler/setup` file, Bundler will automatically check whether it is possible to lock the Bundle. During the lock process, Bundler updates the lock if the implicit locking changes the lock file.

Starting with the 1.10 release, Bundler includes a lockfile section named BUNDLED WITH that includes the version of Bundler that generated the lockfile. In order to minimize git churn, and guarantee that the lockfile will only be changed when the user runs an explicit Bundler command, Bundler will now only add or update the BUNDLED WITH section during commands where the user asks for changes to the lock. This includes, but is not limited to, `install`, `update`, `lock`, and `package`.

Running the `check` command or loading an application that uses Bundler will still now add or update the BUNDLED WITH section if, and only if, the lockfile has also changed for a different reason (such as a gem being updated).

Simply using an application, by running `bundle exec` commands or by running `bin/rails` and the like, will not change the lockfile. As a result, the intended workflow with the BUNDLED WITH section is now slightly different than it was before:

1. When running `bundle install`, Bundler will update the version in the lockfile if newer than the version present.

2. Then, check in the lockfile change, exactly as you would after running install to change any other gem version.

3. Older versions of Bundler will not change the number in the lock, but will warn the user that they are out of date.

refs bundler/bundler-features#80
refs #3697

### CODE REVIEW

**Comment by andremedeiros** [comment]
LGTM. :ship:

File: lib/bundler/definition.rb [hunk 1]
```diff
@@ -658,5 +658,15 @@ def pinned_spec_names(specs)
     def requested_groups
       self.groups - Bundler.settings.without - @optional_groups + Bundler.settings.with
     end
+
+    def lockfiles_equal?(current, proposed, preserve_bundled_with)
+      if preserve_bundled_with
+        pattern = /\n\nBUNDLED WITH\n.*\n/
```
Comment (TimMoore) [inline lib/bundler/definition.rb]: Might be better as a constant.

**Comment by TimMoore** [comment]
Makes sense :+1:

File: lib/bundler/definition.rb (hunk 1 above)
Comment (segiddins) [inline lib/bundler/definition.rb]: use the parser constant

File: spec/commands/check_spec.rb [hunk 2]
```diff
@@ -275,4 +275,59 @@
...
+        bundle :check
+        lockfile_should_be lock_with(nil)
+      end
+    end
+
+    context "is newer" do
+      it "does not change the lock" do
+        lockfile lock_with(Bundler::VERSION.succ)
+        bundle :check
+        lockfile_should_be lock_with(Bundler::VERSION.succ)
```
Comment (segiddins) [inline spec/commands/check_spec.rb]: please verify the warning is still printed

File: spec/other/cli_dispatch_spec.rb [hunk 3]
```diff
@@ -14,7 +14,7 @@
   end
 
   it "print a friendly error when ambiguous" do
-    bundle "i"
+    bundle "in"
```
Comment (segiddins) [inline spec/other/cli_dispatch_spec.rb]: why is this change necessary?

Comment (andremedeiros) [inline spec/other/cli_dispatch_spec.rb]: Because "bundle i " is now mapped to install

File: spec/runtime/setup_spec.rb [hunk 4]
```diff
@@ -899,4 +899,58 @@ def Bundler.require(path)
...
+        ruby "require 'bundler/setup'"
+        lockfile_should_be lock_with(nil)
+      end
+    end
+
+    context "is newer" do
+      it "does not change the lock" do
+        lockfile lock_with(Bundler::VERSION.succ)
+        ruby "require 'bundler/setup'"
+        lockfile_should_be lock_with(Bundler::VERSION.succ)
```
Comment (segiddins) [inline spec/runtime/setup_spec.rb]: check the warning is printed

File: spec/other/cli_dispatch_spec.rb (hunk 3 above)
Comment (TimMoore) [inline spec/other/cli_dispatch_spec.rb]: I'm not sure why it's on this branch, though. I committed it directly to `1-10-stable`. @andremedeiros looks like you cherry-picked the commit?

Comment (indirect) [inline spec/other/cli_dispatch_spec.rb]: I'll just rebase. Please stand by...

File: spec/runtime/setup_spec.rb (hunk 4 above)
Comment (indirect) [inline spec/runtime/setup_spec.rb]: I don't think the warning should be printed. This is the runtime; the user hasn't run a bundler command, bundler is being loaded potentially without the end user's knowledge. I think we should limit warnings to times when the user knows they are interacting with Bundler.

File: spec/commands/check_spec.rb (hunk 2 above)
Comment (indirect) [inline spec/commands/check_spec.rb]: It does at least make sense to me to print the warning here, but I feel like I would rather not warn on `bundle check` than add a third special case to the BUNDLED WITH code :/

File: spec/runtime/setup_spec.rb (hunk 4 above)
Comment (segiddins) [inline spec/runtime/setup_spec.rb]: either way, there should be specs around the warning being printed or not being printed

**Comment by parndt** [comment]
Thanks, @indirect !

File: spec/commands/check_spec.rb (hunk 2 above)
Comment (indirect) [inline spec/commands/check_spec.rb]: Ironic after my last commit, but the existing behavior is my preferred one: the `check` command warns, but `bundler/setup` does not.

**Comment by indirect** [comment]
@segiddins there's your specs

**Comment by indirect** [comment]
Specs are green, so I'm merging this. Thanks for the review, @segiddins!
